c_emulator/rvfi_dii.cpp

```diff
@@ -0,0 +1,249 @@
+   * trace
+   */
+  trace_version = 1;
+  dii_port = port;
+  dii_sock = -1;
+}
+
+uint64_t rvfi_t::get_entry()
+{
+  return 0x80000000;
+}
+
+// returns zero on success
+int rvfi_t::setup_socket(bool config_print)
+{
+  int listen_sock = socket(AF_INET, SOCK_STREAM, 0);
+  if (listen_sock == -1) {
+    fprintf(stderr, "Unable to create socket: %s\n", strerror(errno));
+    return 1;
+  }
+  int reuseaddr = 1;
+  if (setsockopt(listen_sock, SOL_SOCKET, SO_REUSEADDR, &reuseaddr,
+                 sizeof(reuseaddr))
+      == -1) {
+    fprintf(stderr, "Unable to set reuseaddr on socket: %s\n", strerror(errno));
+    return 1;
+  }
+  struct sockaddr_in addr;
+  addr.sin_family = AF_INET;
+  addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+  addr.sin_port = htons(dii_port);
+  if (bind(listen_sock, (struct sockaddr *)&addr, sizeof(addr)) == -1) {
+    fprintf(stderr, "Unable to set bind socket: %s\n", strerror(errno));
+    return 1;
+  }
+  if (listen(listen_sock, 1) == -1) {
+    fprintf(stderr, "Unable to listen on socket: %s\n", strerror(errno));
+    return 1;
+  }
+  socklen_t addrlen = sizeof(addr);
+  if (getsockname(listen_sock, (struct sockaddr *)&addr, &addrlen) == -1) {
+    fprintf(stderr, "Unable to getsockname() on socket: %s\n", strerror(errno));
+    return 1;
+  }
+  printf("Waiting for connection on port %d.\n", ntohs(addr.sin_port));
+  dii_sock = accept(listen_sock, NULL, NULL);
+  if (dii_sock == -1) {
+    fprintf(stderr, "Unable to accept connection on socket: %s\n",
+            strerror(errno));
+    return 1;
+  }
+  close(listen_sock);
+  // Ensure that the socket is blocking
+  int fd_flags = fcntl(dii_sock, F_GETFL);
+  if (fd_flags == -1) {
+    fprintf(stderr, "Failed to get file descriptor flags for socket!\n");
+    return 1;
+  }
+  if (config_print) {
+    fprintf(stderr, "RVFI socket fd flags=%d, nonblocking=%d\n", fd_flags,
+            (fd_flags & O_NONBLOCK) != 0);
+  }
+  if (fd_flags & O_NONBLOCK) {
+    fprintf(stderr, "Socket was non-blocking, this will not work!\n");
+    return 1;
+  }
+  printf("Connected\n");
+  return 0;
+}
+
+void rvfi_t::get_and_send_packet(packet_reader_fn reader, bool config_print)
+{
+  lbits packet;
+  CREATE(lbits)(&packet);
+  reader(&packet, UNIT);
+  /* Note: packet.len is the size in bits, not bytes. */
+  if (packet.len % 8 != 0) {
+    fprintf(stderr, "RVFI-DII trace packet not byte aligned: %d\n",
+            (int)packet.len);
+    exit(1);
+  }
+  const size_t send_size = packet.len / 8;
+  if (config_print) {
+    print_bits("packet = ", packet);
+    fprintf(stderr, "Sending packet with length %zd... ", send_size);
+  }
+  if (send_size > 4096) {
+    fprintf(stderr, "Unexpected large packet size (> 4KB): %zd\n", send_size);
+    exit(1);
+  }
+  unsigned char bytes[send_size];
+  /* mpz_export might not write all of the null bytes */
+  memset(bytes, 0, sizeof(bytes));
+  mpz_export(bytes, NULL, -1, 1, 0, 0, *(packet.bits));
```

Comment:
Now that this code is C++ we could probably move most of the handling out of sail and into the callbacks. But not for this change.

Reply:
I think `AccessModifierOffset: -2` in .clang-format should fix this.